### PR TITLE
Support SETTINGS with implicit ENGINE and mixing engine and query settings

### LINKED ISSUE #63154: CREATE TABLE AS with SETTINGS should recognize settings for the corresponding ENGINE
**Describe the issue**

```
CREATE TABLE hits2 AS hits SETTINGS enable_block_number_column = 1, enable_block_offset_column = 1;
```

Without an explicit ENGINE specification it does not recognize that the settings belong to the MergeTree engine.

### LINKED ISSUE #73099: Cannot create table with SETTINGS with implicit ENGINE
**Describe the issue**

I am trying to create a table with SETTINGS:

This one works:
```
CREATE OR REPLACE TABLE example_table (id UInt32, data String) ENGINE=MergeTree() ORDER BY id SETTINGS max_part_loading_threads=8;
```

but if I omit the ENGINE it fails:
```
CREATE OR REPLACE TABLE example_table (id UInt32, data String) ORDER BY id SETTINGS max_part_loading_threads=8;
...
Received exception from server (version 24.6.1):
Code: 115. DB::Exception: Received from localhost:9000. DB::Exception: Unknown setting 'max_part_loading_threads': Maybe you meant ['max_parsing_threads','max_download_threads']. (UNKNOWN_SETTING)
```

but I can create the table just fine without the SETTINGS:
```
CREATE OR REPLACE TABLE example_table (id UInt32, data String) ORDER BY id;
```

and the engine is defaulted fine
```
show create table example_table;
...
CREATE TABLE default.example_table
(
    `id` UInt32,
    `data` String
)
ENGINE = MergeTree
ORDER BY id
SETTINGS index_granularity = 8192
```

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support SETTINGS with implicit ENGINE and mixing engine and query settings

You can now use engine settings even if the engine is not declared (`default_table_engine`)
```sql
CREATE TABLE example_table
(
    `id` UInt32,
    `data` String
)
ORDER BY id
SETTINGS max_part_loading_threads = 8
```

You can also mix table settings and query settings. The server will apply the server settings dynamically to the query and remove them from the table declaration. In case of ambiguity (the setting can be both table and query setting), it will be kept as a table setting:

```sql
CREATE TABLE example_table
(
    `id` UInt32,
    `data` JSON
)
ENGINE = MergeTree()
ORDER BY id
SETTINGS async_insert = 1, enable_json_type = 1
```

Here `enable_json_type` is applied to the query and the table is created with just `async_insert = 1`

Closes https://github.com/ClickHouse/ClickHouse/issues/73099
Closes https://github.com/ClickHouse/ClickHouse/issues/63154

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 5ecc35d56b56e63ea058ac15e7cd899fd687b252 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/73120/5ecc35d56b56e63ea058ac15e7cd899fd687b252/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/df235c39a782e582c3fc8feec3c8239822e058c1/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by Algunenano** [comment]
Need to disambiguate between engine and query settings

File: src/Interpreters/InterpreterSetQuery.cpp [hunk 1]
```diff
@@ -55,13 +59,40 @@ void InterpreterSetQuery::applySettingsFromQuery(const ASTPtr & ast, ContextMuta
...
+            if (create_query->select)
+                applySettingsFromSelectWithUnion(create_query->select->as<ASTSelectWithUnionQuery &>(), context_);
+            else if (
+                !create_query->settings_ast && create_query->storage && !create_query->storage->engine && create_query->storage->settings
+                && context_->getApplicationType() != Context::ApplicationType::CLIENT)
+            {
+                /// If we parsed one set of settings and no engine we don't know if they belong to the query
+                /// or to the engine. To disambiguate, we try to apply them in the context, if it works then we move
+                /// them there; if it doesn't we assume they are engine settings.
+                /// Note that this only works in the server, since the client ignores unknown settings (for compatibility reasons)
```
Comment (tavplubix) [inline src/Interpreters/InterpreterSetQuery.cpp:78]: There's still some ambiguity:
```
 :) select name from system.settings where name in (select name from system.merge_tree_settings)

SELECT name
FROM system.settings
WHERE name IN (
    SELECT name
    FROM system.merge_tree_settings
)

Query id: 02123176-93ed-44ca-96a6-2dca88094d00

    ┌─name────────────────────────────────────────────────────────┐
 1. │ min_compress_block_size                                     │
 2. │ max_compress_block_size                                     │
 3. │ allow_suspicious_indices                                    │
 4. │ parts_to_delay_insert                                       │
 5. │ parts_to_throw_insert                                       │
 6. │ number_of_mutations_to_delay                                │
 7. │ number_of_mutations_to_throw                                │
 8. │ max_partitions_to_read                                      │
 9. │ async_insert                                                │
10. │ merge_tree_clear_old_temporary_directories_interval_seconds │
11. │ merge_tree_clear_old_parts_interval_seconds                 │
12. │ lightweight_mutation_projection_mode                        │
    └─────────────────────────────────────────────────────────────┘

12 rows in set. Elapsed: 0.017 sec. Processed 1.36 thousand rows, 295.91 KB (82.18 thousand rows/s., 17.91 MB/s.)
Peak memory usage: 7.34 KiB.

```

Comment (Algunenano) [reply src/Interpreters/InterpreterSetQuery.cpp]: I've flipped the checks to remove the ambiguity in a clearer way and support other use cases. Now, if there is only one SETTING declaration (2 of them is unambiguous already) we check all the settings declared to see if they are supported by the storage, and if not check if they are global settings to move them there.

This supports mixing the 2 types of settings together: e.g `ENGINE MergeTree() ORDER BY tuple() SETTINGS allow_experimental_json_type=1, replace_long_file_name_to_hash=1` will move the JSON setting to the context (and allow creating a table with JSON in that statement) while keeping the MergeTree setting in the definition.

I'm currently running CI to see what breaks and will add more tests to verify things work correctly

File: src/Storages/StorageFactory.h [hunk 2]
```diff
@@ -29,6 +29,10 @@ class StorageFactory : private boost::noncopyable, public IHints<>
 
     static StorageFactory & instance();
 
+    /// Helper function to validate if a specific storage supports a setting
+    /// Meant to validate if a table engine before query start
```
Comment (Avogar) [inline src/Storages/StorageFactory.h:33]: What did you mean by `validate if a table engine before query start` here?

Comment (Algunenano) [reply src/Storages/StorageFactory.h:33]: I'll reword this properly. It's used to validate if settings are query or engine settings before start the query interpretation

**Review by Avogar — APPROVED** [review]

**Comment by clickgapai** [comment]
Hi @Algunenano @Avogar — while reviewing this PR I found the following:

- **Bug (P2):** `ATTACH TABLE ... SETTINGS` silently ignores settings when ENGINE not specified — https://github.com/ClickHouse/ClickHouse/issues/104791

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**